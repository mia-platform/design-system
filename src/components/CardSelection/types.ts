
export enum Type {
  Radio = 'radio',
  Checkbox = 'checkbox'
}
