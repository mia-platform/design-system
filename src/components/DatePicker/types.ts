import {
  DisabledDate as AntDisabledDate,
} from 'rc-picker/lib/interface'
import { Dayjs } from 'dayjs'

export type DisabledDate = AntDisabledDate<Dayjs>
export enum ShowTimeOptions {
  Hours = 'hours',
  Minutes = 'minutes',
  Seconds = 'seconds'
}

export type ShowTime = boolean | {
  showHour: boolean;
  showMinute: boolean;
  showSecond: boolean;
}
