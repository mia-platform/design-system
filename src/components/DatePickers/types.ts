/**
 * Copyright 2025 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  DisabledDate as AntDisabledDate,
} from 'rc-picker/lib/interface'
import { Dayjs } from 'dayjs'

export type DisabledDate = AntDisabledDate<Dayjs>
export enum ShowTimeOptions {
  Hours = 'hours',
  Minutes = 'minutes',
  Seconds = 'seconds'
}

export type AntShowTimeOptions = boolean | {
  showHour: boolean;
  showMinute: boolean;
  showSecond: boolean;
}

export type NoUndefinedRangeValueType = [start: Dayjs | null, end: Dayjs | null];

export type RangeValueType = [
  start: Dayjs | null | undefined,
  end: Dayjs | null | undefined
];

export const defaultTimeFormat = 'HH:mm'
export const defaultDateFormat = 'DD/MM/YYYY'
