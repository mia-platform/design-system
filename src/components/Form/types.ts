
export enum Layout {
  Horizontal = 'horizontal',
  Vertical = 'vertical'
}
