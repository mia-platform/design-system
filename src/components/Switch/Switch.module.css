/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

.switchComponent {
  display: flex;
  flex-direction: column;
  gap: var(--spacing-gap-xs, 4px);

  .switchTextWrapper {
    display: flex;
    align-items: center;
    gap: var(--spacing-gap-md, 12px);
    color: var(--palette-text-neutral-main, #636363);
  }

  .switchDescription {
    padding-right: var(--spacing-padding-sm, 8px);
    padding-left: calc(var(--spacing-padding-sm, 8px) + var(--spacing-padding-3xl, 48px));
    color: var(--palette-text-neutral-main, #636363);

    &.small {
      padding-left: calc(var(--spacing-padding-sm, 8px) + var(--spacing-padding-2xl, 32px));
    }
  }

  :global(.mia-platform-switch) {
    background: var(--palette-action-secondary-bold, #cdcdcd);

    &:hover {
       background: var(--palette-action-secondary-bolder, #acacac);
    }

    /* Disabled or loading switch */
    
    &:global(.mia-platform-switch-disabled), &:global(.mia-platform-switch-loading) {
      opacity: 1;
      background: var(--palette-action-disabled-main, #f2f2f2);

      &:hover {
        background: var(--palette-action-disabled-main, #f2f2f2);
      }

      :global(.mia-platform-switch-loading-icon) {
        color: var(--palette-action-disabled-main, #f2f2f2);
      }
    }

    /* Checked switch */

    &:global(.mia-platform-switch-checked) {
      background: var(--palette-action-primary-default, #1261e4);

      &:hover {
        background: var(--palette-action-primary-hover, #1890ff);
      }

      /* Disabled or loading checked switch */

      &:global(.mia-platform-switch-disabled), &:global(.mia-platform-switch-loading){
        background: var(--palette-background-primary-0, #aad1fb);

        &:hover {
          background: var(--palette-background-primary-0, #aad1fb);
        }

        :global(.mia-platform-switch-loading-icon) {
          color: var(--palette-background-primary-0, #aad1fb);
        }
      }
    }

    /* Switch handle */

    :global(.mia-platform-switch-handle::before) {
      background: var(--palette-action-secondary-main, #fff);
    }
  }
}
