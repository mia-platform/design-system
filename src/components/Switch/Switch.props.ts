/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { SwitchChangeEventHandler, SwitchClickEventHandler } from 'antd/es/switch'

import { Size } from './Switch.types'

export type SwitchProps = {

    /**
     * Determines whether the switch is checked.
     */
    isChecked?: boolean,

    /**
     * Determines whether the switch is disabled. Defaults to false.
     */
    isDisabled?: boolean,

    /**
     * Determines whether the switch is checked on its first render.
     * It is overridden by the isChecked prop, when present.
     * Defaults to false.
     */
    isInitiallyChecked?: boolean,

    /**
     * Determines whether the switch is in loading state. Defaults to false.
     */
    isLoading?: boolean,

    /**
    * Function that is invoked when the switch state is changed.
    */
    onChange?: SwitchChangeEventHandler

    /**
    * Function that is invoked when the switch is clicked.
    */
    onClick?: SwitchClickEventHandler,

    /**
     * Determines the switch size. Can be set to `large` or `small`. Defaults to `large`.
     */
    size?: Size,
}
