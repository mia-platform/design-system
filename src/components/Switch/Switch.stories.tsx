/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'

import { Size } from './Switch.types'
import { Switch } from '.'
import { defaults } from './Switch'

// NOTE: we are not using the `actions` addon here to simulate the onClick and onChange props
// because it is causing glitches in the switch animation.
// We are reproducing the addon's behaviour by logging the function parameters in the dev console.

const meta = {
  component: Switch,
  args: {
    ...defaults,
    // eslint-disable-next-line no-console
    onChange: (...params) => { console.log('onChange params: ', params) },
    // eslint-disable-next-line no-console
    onClick: (...params) => { console.log('onClick params: ', params) },
  },
  argTypes: {
    description: { type: 'string' },
    text: { type: 'string' },
    onClick: { control: false },
    onChange: { control: false },
  },
} satisfies Meta<typeof Switch>

export default meta
type Story = StoryObj<typeof meta>

export const BasicExample: Story = {}

export const CheckedOnFirstRender: Story = {
  args: {
    isInitiallyChecked: true,
  },
}

export const ControlledCheckedState: Story = {
  args: {
    isChecked: true,
  },
}

export const Disabled: Story = {
  args: {
    isDisabled: true,
  },
}

export const CheckedDisabled: Story = {
  args: {
    isChecked: true,
    isDisabled: true,
  },
}

export const Loading: Story = {
  args: {
    isLoading: true,
  },
}

export const CheckedLoading: Story = {
  args: {
    isChecked: true,
    isLoading: true,
  },
}

export const WithText: Story = {
  args: {
    text: 'Switch with text',
  },
}

export const WithTextAndDescription: Story = {
  args: {
    text: 'Switch with text',
    description: 'This is a description of the switch.',
  },
}

export const SmallSize: Story = {
  args: {
    size: Size.Small,
  },
}

export const SmallSizeWithText: Story = {
  args: {
    size: Size.Small,
    text: 'Switch with text',
  },
}

export const SmallSizeWithTextAndDescription: Story = {
  args: {
    size: Size.Small,
    text: 'Switch with text',
    description: 'This is a description of the switch.',
  },
}

export const WithOnChange: Story = {
  args: {
    onChange: (checked) => alert(`Switch is now ${checked ? 'checked' : 'unchecked'}`),
  },
}

export const WithOnClick: Story = {
  args: {
    onClick: () => alert('Switch has been clicked'),
  },
}
