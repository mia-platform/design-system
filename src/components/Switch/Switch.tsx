/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactElement, useMemo } from 'react'
import { Switch as AntSwitch } from 'antd'
import type { SwitchSize } from 'antd/es/switch'
import classNames from 'classnames'

import { BodyS } from '../Typography/BodyX/BodyS'
import { Size } from './Switch.types'
import { SwitchProps } from './Switch.props'
import styles from './Switch.module.css'

const {
  switchComponent,
  switchTextWrapper,
  switchDescription,
  small: smallSwitch,
} = styles

export const defaults = {
  isDisabled: false,
  isInitiallyChecked: false,
  isLoading: false,
  size: Size.Large,
}

const antSizeRemapping = {
  [Size.Large]: 'default' as SwitchSize,
  [Size.Small]: 'small' as SwitchSize,
}

export const Switch = ({
  description,
  isChecked,
  isInitiallyChecked = defaults.isInitiallyChecked,
  isDisabled = defaults.isDisabled,
  isLoading = defaults.isLoading,
  onChange,
  onClick,
  size = defaults.size,
  text,
} : SwitchProps) : ReactElement => {
  const descriptionClassName = useMemo(() => classNames([
    switchDescription,
    size === Size.Small && smallSwitch,
  ]), [size])

  return (
    <div className={switchComponent}>
      <div className={switchTextWrapper}>
        <AntSwitch
          checked={isChecked}
          defaultChecked={isInitiallyChecked}
          disabled={isDisabled}
          loading={isLoading}
          size={antSizeRemapping[size]}
          onChange={onChange}
          onClick={onClick}
        />
        {text && <BodyS>{text}</BodyS>}
      </div>
      {
        text && description && (
          <div className={descriptionClassName}>
            <BodyS>{description}</BodyS>
          </div>
        )
      }
    </div>
  )
}

Switch.Size = Size
