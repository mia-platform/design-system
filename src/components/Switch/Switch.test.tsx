import { render, screen, userEvent, waitFor } from '../../test-utils'
import { Size } from './Switch.types'
import { Switch } from './Switch'

describe('Switch', () => {
  describe('renders states correctly', () => {
    it('renders default state', () => {
      render(<Switch />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).not.toBeChecked()
    })

    it('renders checked by default state', () => {
      render(<Switch isInitiallyChecked={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeChecked()
    })

    it('renders checked state', () => {
      render(<Switch isChecked={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeChecked()
    })

    it('renders default disabled state', () => {
      render(<Switch isDisabled={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeDisabled()
    })

    it('renders checked disabled state', () => {
      const { asFragment } = render(<Switch isChecked={true} isDisabled={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeDisabled()
      expect(screen.getByRole('switch')).toBeChecked()
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders default loading state', () => {
      const { asFragment } = render(<Switch isChecked={false} isLoading={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeDisabled()
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders checked loading state', () => {
      const { asFragment } = render(<Switch isChecked={true} isLoading={true} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toBeDisabled()
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('renders sizes correctly', () => {
    it('renders large size', () => {
      render(<Switch size={Size.Large} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toHaveClass('mia-platform-switch')
      expect(screen.getByRole('switch')).not.toHaveClass('mia-platform-switch-small')
    })

    it('renders small size', () => {
      render(<Switch size={Size.Small} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).toHaveClass('mia-platform-switch')
      expect(screen.getByRole('switch')).toHaveClass('mia-platform-switch-small')
    })
  })

  describe('renders text correctly', () => {
    it('renders simple text', () => {
      render(<Switch text="Switch" />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByText('Switch')).toBeInTheDocument()
    })

    it('renders text and description', () => {
      render(<Switch description="Switch description" text="Switch text" />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByText('Switch text')).toBeInTheDocument()
      expect(screen.getByText('Switch description')).toBeInTheDocument()
    })

    it('ignores description if text is not set', () => {
      render(<Switch description="Switch description" />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.queryByText('Switch description')).not.toBeInTheDocument()
    })
  })

  describe('performs interactions correctly', () => {
    it('calls onClick when the switch is clicked', async() => {
      const onClick = jest.fn()
      render(<Switch onClick={onClick} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).not.toBeChecked()
      userEvent.click(screen.getByRole('switch'))
      await waitFor(() => expect(onClick).toHaveBeenCalledTimes(1))
    })

    it('calls onChange when the switch state changes', async() => {
      const onChange = jest.fn()
      render(<Switch onChange={onChange} />)
      expect(screen.getByRole('switch')).toBeInTheDocument()
      expect(screen.getByRole('switch')).not.toBeChecked()
      userEvent.click(screen.getByRole('switch'))
      await waitFor(() => expect(onChange).toHaveBeenCalledTimes(1))
    })
  })
})
