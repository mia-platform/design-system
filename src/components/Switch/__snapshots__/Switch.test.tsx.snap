// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders states correctly renders checked disabled state 1`] = `
<DocumentFragment>
  <div
    class="switchComponent"
  >
    <div
      class="switchTextWrapper"
    >
      <button
        aria-checked="true"
        class="mia-platform-switch mia-platform-switch-checked mia-platform-switch-disabled"
        disabled=""
        role="switch"
        type="button"
      >
        <div
          class="mia-platform-switch-handle"
        />
        <span
          class="mia-platform-switch-inner"
        >
          <span
            class="mia-platform-switch-inner-checked"
          />
          <span
            class="mia-platform-switch-inner-unchecked"
          />
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Switch renders states correctly renders checked loading state 1`] = `
<DocumentFragment>
  <div
    class="switchComponent"
  >
    <div
      class="switchTextWrapper"
    >
      <button
        aria-checked="true"
        class="mia-platform-switch mia-platform-switch-loading mia-platform-switch-checked mia-platform-switch-disabled"
        disabled=""
        role="switch"
        type="button"
      >
        <div
          class="mia-platform-switch-handle"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin mia-platform-switch-loading-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </div>
        <span
          class="mia-platform-switch-inner"
        >
          <span
            class="mia-platform-switch-inner-checked"
          />
          <span
            class="mia-platform-switch-inner-unchecked"
          />
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Switch renders states correctly renders default loading state 1`] = `
<DocumentFragment>
  <div
    class="switchComponent"
  >
    <div
      class="switchTextWrapper"
    >
      <button
        aria-checked="false"
        class="mia-platform-switch mia-platform-switch-loading mia-platform-switch-disabled"
        disabled=""
        role="switch"
        type="button"
      >
        <div
          class="mia-platform-switch-handle"
        >
          <span
            aria-label="loading"
            class="anticon anticon-loading anticon-spin mia-platform-switch-loading-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="loading"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
              />
            </svg>
          </span>
        </div>
        <span
          class="mia-platform-switch-inner"
        >
          <span
            class="mia-platform-switch-inner-checked"
          />
          <span
            class="mia-platform-switch-inner-unchecked"
          />
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
